src/libmeasurement_kit/engine.cpp

```diff
@@ -0,0 +1,480 @@
+#include <atomic>
+#include <condition_variable>
+#include <deque>
+#include <exception>
+#include <future>
+#include <memory>
+#include <mutex>
+#include <set>
+#include <string>
+#include <sstream>
+#include <thread>
+#include <tuple>
+#include <type_traits>
+
+#include <measurement_kit/common/error.hpp>
+#include <measurement_kit/common/logger.hpp>
+#include <measurement_kit/common/nlohmann/json.hpp>
+#include <measurement_kit/common/reactor.hpp>
+#include <measurement_kit/common/shared_ptr.hpp>
+
+#include "src/libmeasurement_kit/nettests/runnable.hpp"
+
+namespace mk {
+namespace engine {
+
+// # Multi-thread stuff
+//
+// Comes first because it needs more careful handling.
+
+class Semaphore {
+  public:
+    Semaphore() = default;
+
+    void wait() {
+        std::unique_lock<std::mutex> lock{mutex_};
+        cond_.wait(lock, [this]() { return !active_; });
+        active_ = true;
+    }
+
+    void signal() {
+        {
+            std::unique_lock<std::mutex> _{mutex_};
+            active_ = false;
+        }
+        cond_.notify_one();
+    }
+
+    ~Semaphore() = default;
+
+  private:
+    bool active_ = false;
+    std::condition_variable cond_;
+    std::mutex mutex_;
+};
+
+class TaskImpl {
+  public:
+    std::condition_variable cond;
+    std::deque<nlohmann::json> deque;
+    std::atomic_bool interrupted{false};
+    std::mutex mutex;
+    SharedPtr<Reactor> reactor = Reactor::make();
+    std::atomic_bool running{false};
+    std::thread thread;
+};
+
+static void task_run(TaskImpl *pimpl, nlohmann::json &settings);
+static bool is_event_valid(const std::string &);
+
+static void emit(TaskImpl *pimpl, nlohmann::json &&event) {
+
+    // In debug mode, make sure that we're emitting an event that we know
+    assert(event.is_object());
+    assert(event.count("type") != 0);
+    assert(event.at("type").is_string());
+    assert(is_event_valid(event.at("type").get<std::string>()));
+
+    // Actually emit the event.
+    {
+        std::unique_lock<std::mutex> _{pimpl->mutex};
+        pimpl->deque.push_back(std::move(event));
+    }
+    pimpl->cond.notify_all(); // more efficient if unlocked
+}
+
+Task::Task(nlohmann::json &&settings) {
+    pimpl_ = std::make_unique<TaskImpl>();
+    std::promise<void> barrier;
+    std::future<void> started = barrier.get_future();
+    pimpl_->thread = std::thread([this, &barrier,
+                                         settings = std::move(
+                                                 settings)]() mutable {
+        pimpl_->running = true;
+        barrier.set_value();
+        static Semaphore semaphore;
+        semaphore.wait(); // block until a previous task has finished running
+        task_run(pimpl_.get(), settings);
+        pimpl_->running = false;
+        pimpl_->cond.notify_all(); // tell the readers we're done
+        semaphore.signal();        // allow another task to run
+    });
+    started.wait(); // guarantee Task() completes when the thread is running
```

Comment:
See my bikeshed comment above, but I think it's a bit confusing to have the semaphore and a `future` have similar interface. I had to grep through the code to be sure this was not another semaphore.

Reply:
I didn't consider this. I think changing the names of the `Semaphore` can make the code more easy to read, so I will do that. Thanks for pointing out this.